### ISSUE TITLE
Creating an LIst or Ensemble of MLPs, then evaluating on a different subset each iteration?

### ISSUE THREAD

**Issue by adam-hartshorne** [patrick-kidger/equinox#905]
I wish to create N independent MLPs, where N is determined at runtime. Then each iteration, I wish to evaluate a subset of them on different data where the indices of the MLPs  (if we imagine they were stored in a list) I wish to evaluate is passed as part of the data.

AFAIK, you can't create a List or Tuple of MLPs then evaluate each one using subscripting as the batch of ids is a traced array.

From https://docs.kidger.site/equinox/tricks/ , I understand how to create an ensemble of MLPs, but this example only describes how to evaluate inputs on all MLPs in the ensemble. I could do this each time and mask out the results I wish to know, however this will get very inefficient if N is large.

I was wondering if anybody had any suggests how to achieve this.

**Comment by lockwo** [comment]
One approach would be to just collect those indices that you need, e.g. something like

```python
import jax
from jax import numpy as jnp
import equinox as eqx

key = jax.random.PRNGKey(0)
keys = jax.random.split(key, 8)

# Create an ensemble of models
@eqx.filter_vmap
def make_ensemble(key):
    return eqx.nn.MLP(2, 2, 2, 2, key=key)

mlp_ensemble = make_ensemble(keys)

@eqx.filter_jit
def evaluate_per_inds(model, x, inds):
    arrs, static = eqx.partition(model, eqx.is_array)
    subset_model = eqx.combine(jax.tree.map(lambda x: x[inds], arrs), static)
    return eqx.filter_vmap(lambda a, b: a(b), in_axes=(eqx.if_array(0), 0))(subset_model, x)

evaluate_per_inds(mlp_ensemble, jax.random.normal(key, (4, 2)), jnp.array([0, 2, 4, 5]))
```

**Comment by adam-hartshorne** [comment]
Perfect. Thanks very much.
